### PR TITLE
Add Config Registry for autotune

### PR DESCRIPTION
**Pull request by Kaihui-intel**
## Type of Change

feature

## Description

https://github.com/intel/neural-compressor/discussions/1445#discussioncomment-8128432

```diff
- from neural_compressor.common.base_config import registered_configs
+ from neural_compressor.common.base_config import config_registry
```
@mengniwang95, @xin3he @zehao-intel Please be aware of the change.

## Expected Behavior & Potential Risk

the expected behavior that triggered by this PR 

## How has this PR been tested?

Pre-CI

## Dependency Change?

None

### CODE REVIEW

File: neural_compressor/common/base_config.py [hunk 1]
```diff
@@ -40,30 +40,71 @@
...
 # Dictionary to store registered configurations
-registered_configs = {}
 
 
-def register_config(framework_name="None", algo_name=None):
+class ConfigRegistry:
+    registered_configs = {}
+
+    @classmethod
+    def register_config(cls, framework_name="None", algo_name=None, priority=0):
```
Comment (yiliu30) [inline neural_compressor/common/base_config.py:49]: How about renaming the register_config to register_config_impl to distinguish the API and implementation?
```python
@register_config(...)  # API
Class XxxConfig:
      ....

```

File: neural_compressor/common/base_config.py [hunk 2]
```diff
@@ -40,30 +40,71 @@
...
+
+        Usage example:
+            @ConfigRegistry.register_config(framework_name=FRAMEWORK_NAME, algo_name=ExampleAlgorithm, priority=100)
+            class ExampleAlgorithmConfig:
+                # Configuration details for the ExampleAlgorithm
+
+        Args:
+            framework_name: the framework name. Defaults to "None".
+            algo_name: the algorithm name. Defaults to None.
+            priority: the priority of the configuration. Defaults to 0.
```
Comment (yiliu30) [inline neural_compressor/common/base_config.py:62]: Please add the explanation about priority here.

File: neural_compressor/common/base_config.py [hunk 3]
```diff
@@ -337,10 +378,10 @@ def from_dict(cls, config_dict: OrderedDict[str, Dict], config_registry: Dict[st
         assert len(config_dict) >= 1, "The config dict must include at least one configuration."
         num_configs = len(config_dict)
         name, value = next(iter(config_dict.items()))
-        config = config_registry[name].from_dict(value)
+        config = config_registry[name]["cls"].from_dict(value)
```
Comment (yiliu30) [inline neural_compressor/common/base_config.py:381]: How about adding a method to `ConfigRegistry` to get the `cls`, like `get_cls_by_name(name: str) -> BaseConfig`?

File: neural_compressor/torch/utils/constants.py [hunk 4]
```diff
@@ -36,3 +36,7 @@
         "double_quant_group_size": 256,
     },
 }
+
+# Setting priorities for algorithms, a higher number indicates a higher priority.
```
Comment (yiliu30) [inline neural_compressor/torch/utils/constants.py:40]:
```suggestion
# Setting priorities for algorithms, a larger number indicates a higher priority, which will be tried first at the auto-tune stage.
```

**Review by yiliu30 — COMMENTED** [review]
Another suggestion, please add the typing hints as much as possible.

**Review by yiliu30 — APPROVED** [review]

**Review by zehao-intel — APPROVED** [review]

**Review by mengniwang95 — APPROVED** [review]

**Pull request merged**